run-tests.php

```diff
@@ -1645,11 +1645,47 @@ function run_all_tests_parallel(array $test_files, array $env, $redir_tested): v
     }
 }
 
+/**
+ * Calls fwrite and retries when network writes fail with errors such as "Resource temporarily unavailable"
+ *
+ * @param resource $stream the stream to fwrite to
+ * @param string $data
+ * @return int|false
+ */
+function safe_fwrite($stream, string $data)
+{
+    // safe_fwrite was tested by adding $message['unused'] = str_repeat('a', 20_000_000); in send_message()
+    // fwrites on tcp sockets can return false or less than strlen if the recipient is busy.
+    // (e.g. fwrite(): Send of 577 bytes failed with errno=35 Resource temporarily unavailable)
+    $bytes_written = 0;
+    while ($bytes_written < strlen($data)) {
+        $n = @fwrite($stream, substr($data, $bytes_written));
+        if ($n === false) {
+            $write_streams = [$stream];
+            $read_streams = [];
+            $except_streams = [];
+            /* Wait for up to 10 seconds for the stream to be ready to write again. */
+            $result = stream_select($read_streams, $write_streams, $except_streams, 10);
+            if (!$result) {
+                echo "ERROR: send_message() stream_select() failed\n";
```

Comment:
```suggestion
                fwrite(STDERR, "ERROR: send_message() stream_select() failed\n");
```

Reply:
```php
function error(string $message): void
{
    echo "ERROR: {$message}\n";
    exit(1);
}
```

Nothing else in run-tests.php is printing to STDERR or referencing STDERR yet - starting to do that would introduce an inconsistency with similar error logging.

For anything redirecting to a log file, the existing behavior of printing to stdout seems fine, especially since there's no machine-readable output format for run-tests.php as an option